.gitmodules

```diff
@@ -22,3 +22,6 @@
 [submodule "tools/jamulus-server-remote"]
 	path = tools/jamulus-server-remote
 	url = https://github.com/vdellamea/jamulus-server-remote
+[submodule "libs/portaudio"]
+	path = libs/portaudio
+	url = https://github.com/npostavs/portaudio.git
```

Comment:
You have referenced your fork. I would like to have to original repo referenced here.

Reply:
The repo commit you have used is 3988fd. Is this an official release version of the portaudio library?

Reply:
> The repo commit you have used is 3988fd. Is this an official release version of the portaudio library?

No, I added a couple of commits to last stable release (hence why I used my fork instead of upstream):
1. [Backport to let the wasapi code build](https://github.com/PortAudio/portaudio/commit/d7e79160a3670dc89ad35417ea511d9196964927)
2. [Modification to handle ASIO buffer size changes](https://github.com/PortAudio/portaudio/commit/3988fd9ebc10cd7b1834ec6e0c0c01c9f6dbf27a)

Reply:
Can you please create pull requests for your changes in the official repo?

Reply:
The first is already from portaudio's master branch. I've just made a pull request for the second one: PortAudio/portaudio#473

Reply:
I can see you've not had much progress over on the PortAudio PR.  Had any luck finding out about the thread safety issues?

Reply:
> Had any luck finding out about the thread safety issues?

No. I guess the fact that Jamulus' current Windows code has `ASIOMutex` is suggestive, although it's not used for `CSound::LoadAndInitializeDriver`(!?)

Reply:
OK, quick review of the threading model from memory - I've not looked at it closely on the client:
- the client main thread starts (single thread)
- if there's an audio driver, the audio thread should start (single thread)
- the network thread starts (single thread)
- if there's a GUI, the GUI thread should start (single thread)

I _think_ that's it.  I may be mistaken about the separate audio thread - it could be the main thread (which would otherwise only be co-ordinating network and audio and responding to and informing the GUI).

Where would multi-threading cause contention in audio driver processing in the client?  I'd guess it can't before the driver is running, at least.

Reply:
So if the device's buffer size gets changed, it sends a `kAsioResetRequest` message in the audio thread. Jamulus' handler calls `SoundBase::EmitReinitRequestSignal` which does `emit ReinitRequest`. That's connected to the `CClient::OnSndCrdReinitRequest` slot, which would run in the GUI thread?